lib/validation.js

```diff
@@ -1,40 +1,36 @@
 'use strict'
 
-const fastJsonStringify = require('fast-json-stringify')
-const Ajv = require('ajv')
-
-const bodySchema = Symbol('body-schema')
-const querystringSchema = Symbol('querystring-schema')
+const headersSchema = Symbol('headers-schema')
 const paramsSchema = Symbol('params-schema')
+const querystringSchema = Symbol('querystring-schema')
+const bodySchema = Symbol('body-schema')
+
 const responseSchema = Symbol('response-schema')
-const headersSchema = Symbol('headers-schema')
-const kFluentSchema = Symbol.for('fluent-schema-object')
 
-function getValidatorForStatusCodeSchema (statusCodeDefinition, externalSchema) {
-  return fastJsonStringify(statusCodeDefinition, { schema: externalSchema })
-}
+function compileSchemasForSerialization (context, compile) {
+  if (!context.schema || !context.schema.response) {
+    return
+  }
 
-function getResponseSchema (responseSchemaDefinition, sharedSchemas) {
-  var statusCodes = Object.keys(responseSchemaDefinition)
-  return statusCodes.reduce(function (r, statusCode) {
-    r[statusCode] = getValidatorForStatusCodeSchema(responseSchemaDefinition[statusCode], sharedSchemas)
-    return r
-  }, {})
+  const { method, url } = context.config || {}
+  context[responseSchema] = Object.keys(context.schema.response)
+    .reduce(function (acc, statusCode) {
+      acc[statusCode] = compile(method, url, statusCode, context.schema.response[statusCode])
+      return acc
+    }, {})
 }
 
-function build (context, compile, schemas, schemaResolver) {
+function compileSchemasForValidation (context, compile) {
   if (!context.schema) {
     return
   }
 
-  generateFluentSchema(context.schema)
-  context.schema = schemas.resolveRefs(context.schema, false, schemaResolver)
+  const { method, url } = context.config || {}
 
   const headers = context.schema.headers
-
   if (headers && Object.getPrototypeOf(headers) !== Object.prototype) {
     // do not mess with non-literals, e.g. Joi schemas
-    context[headersSchema] = compile(headers)
+    context[headersSchema] = compile(method, url, 'headers', headers)
```

Comment:
Why does this function incur side effects?

Reply:
As in the [other comment](https://github.com/fastify/fastify/pull/2023#discussion_r372045347), I would keep only one thread 😄